core/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java

```diff
@@ -61,84 +58,55 @@ public TransportRecoveryAction(Settings settings, ThreadPool threadPool, Cluster
                                    TransportService transportService, IndicesService indicesService,
                                    ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
         super(settings, RecoveryAction.NAME, threadPool, clusterService, transportService, actionFilters, indexNameExpressionResolver,
-                RecoveryRequest.class, ShardRecoveryRequest.class, ThreadPool.Names.MANAGEMENT);
+                RecoveryRequest.class, ThreadPool.Names.MANAGEMENT);
         this.indicesService = indicesService;
     }
 
     @Override
-    protected RecoveryResponse newResponse(RecoveryRequest request, AtomicReferenceArray shardsResponses, ClusterState clusterState) {
+    protected RecoveryState readShardResult(StreamInput in) throws IOException {
+        return RecoveryState.readRecoveryState(in);
+    }
 
-        int successfulShards = 0;
-        int failedShards = 0;
-        List<ShardOperationFailedException> shardFailures = null;
-        Map<String, List<ShardRecoveryResponse>> shardResponses = new HashMap<>();
 
-        for (int i = 0; i < shardsResponses.length(); i++) {
-            Object shardResponse = shardsResponses.get(i);
-            if (shardResponse == null) {
-                // simply ignore non active shards
-            } else if (shardResponse instanceof BroadcastShardOperationFailedException) {
-                failedShards++;
-                if (shardFailures == null) {
-                    shardFailures = new ArrayList<>();
+    @Override
+    protected RecoveryResponse newResponse(RecoveryRequest request, int totalShards, int successfulShards, int failedShards, List<RecoveryState> responses, List<ShardOperationFailedException> shardFailures) {
+        Map<String, List<RecoveryState>> shardResponses = Maps.newHashMap();
```

Comment:
kk. was referring to both the maps and the lists later on…

> On 28 Aug 2015, at 20:40, Jason Tedor <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java:
> 
> >  
> > -        for (int i = 0; i < shardsResponses.length(); i++) {
> > -            Object shardResponse = shardsResponses.get(i);
> > -            if (shardResponse == null) {
> > -                // simply ignore non active shards
> > -            } else if (shardResponse instanceof BroadcastShardOperationFailedException) {
> > -                failedShards++;
> > -                if (shardFailures == null) {
> > -                    shardFailures = new ArrayList<>();
> > +    @Override
> > +    protected RecoveryResponse newResponse(RecoveryRequest request, int totalShards, int successfulShards, int failedShards, List<RecoveryState> responses, List<ShardOperationFailedException> shardFailures) {
> > +        Map<String, List<RecoveryState>> shardResponses = Maps.newHashMap();
> 
> @bleskes Are you referring to Maps? That hasn't been forbidden yet (but it will be soon).
> 
> —
> Reply to this email directly or view it on GitHub.
>